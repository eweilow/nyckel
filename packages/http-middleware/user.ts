import {
  attemptToRefreshToken,
  GlobalAuthenticationConfig,
  getUserInfo,
  decodeJWT,
  DecodedAccessToken
} from "@nyckel/authentication";
import { SessionManager } from "@nyckel/sessions";
import {
  getManagementUserInfo,
  getManagementUserRoles,
  getManagementUserPermissions
} from "@nyckel/management";

import { SessionInRequest } from "./session";
import { DecodedIdToken } from "@nyckel/authentication/jwt/verify";

export type UserInRequest = {
  get(): Promise<
    | ({
        accessTokenData: DecodedAccessToken;
        idTokenData: DecodedIdToken;
        accessToken: string;
        idToken: string;
        expiresIn: number;
      })
    | null
  >;
  userInfo(): Promise<any>;
  fullUserData(): Promise<any>;
  roles(): Promise<any>;
  permissions(): Promise<any>;
};

function isExpired(
  expires: number,
  now: number = Date.now(),
  margin: number = 10 * 60 * 1000 // Refreshing 10 minutes before token expires give ample time for multiple retries
) {
  return now >= expires - margin;
}

export { isExpired as isTokenExpired };

export function enhanceRequestWithUser<T extends { session: SessionInRequest }>(
  request: T,
  cookieName: string,
  cookie: string,
  sessions: SessionManager,
  authConfig: GlobalAuthenticationConfig,
  authManagementConfig: GlobalAuthenticationConfig,
  expirationMargin: number = 10 * 60 * 1000
): T & { user: UserInRequest } {
  let enhancedRequest = request as T & { user: UserInRequest };

  enhancedRequest.user = {
    async get() {
      const id = cookie;
      if (!sessions.isValidId(id)) {
        return null;
      }

      let session = await enhancedRequest.session.get();
      if (session == null) {
        return null;
      }
      if (session.accessToken == null) {
        return null;
      }

      if (isExpired(session.expires, Date.now(), expirationMargin)) {
        const lock = await sessions.lock(id, 5000);
        if (lock.acquired) {
          try {
            const refreshed = await attemptToRefreshToken(session, authConfig);
            session = {
              ...session,
              ...refreshed
            };
            await sessions.set(id, session);
          } finally {
            await lock.release();
          }
        }
      }

      return {
        idTokenData: decodeJWT(session.idToken),
        accessTokenData: decodeJWT(session.accessToken),
        accessToken: session.accessToken,
        idToken: session.idToken,
        expiresIn: session.expires - Date.now()
      };
    },
    async userInfo() {
      const user = await enhancedRequest.user.get();
      if (user == null) {
        return null;
      }
      return getUserInfo(user.accessToken, authConfig);
    },
    async fullUserData() {
      const user = await enhancedRequest.user.get();
      if (user == null) {
        return null;
      }
      return getManagementUserInfo(user.idTokenData.sub!, authManagementConfig);
    },
    async roles() {
      const user = await enhancedRequest.user.get();
      if (user == null) {
        return null;
      }
      return getManagementUserRoles(
        user.idTokenData.sub!,
        authManagementConfig
      );
    },
    async permissions() {
      const user = await enhancedRequest.user.get();
      if (user == null) {
        return null;
      }
      return getManagementUserPermissions(
        user.idTokenData.sub!,
        authManagementConfig
      );
    }
  };

  return enhancedRequest;
}
