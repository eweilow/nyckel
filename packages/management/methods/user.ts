import {
  GlobalAuthenticationConfig,
  concatUrl,
  verifyAPIResponse
} from "@nyckel/authentication";
import { managementRateLimiter } from "./limiter";
import fetch from "node-fetch";
import { getManagementToken } from "../token";

export async function getManagementUserInfo(
  id: string,
  config: GlobalAuthenticationConfig
) {
  await managementRateLimiter.wait("*");

  const url = concatUrl(
    config.authorizationDomain,
    "/api/v2/users/" + encodeURIComponent(id)
  );
  const token = await getManagementToken(config);
  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${token}`
    }
  });

  managementRateLimiter.updateFromResponse("*", response);
  if (response.status === 429) {
    return getManagementUserInfo(id, config);
  }

  const json = await response.json();

  return verifyAPIResponse(
    response.status,
    json,
    concatUrl(config.authorizationDomain, "/api/v2/users/<id>")
  );
}
