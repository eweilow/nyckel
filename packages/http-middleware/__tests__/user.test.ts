import { enhanceRequestWithUser } from "../user";

describe("enhanceRequestWithUser", () => {
  it("should enhance a request correctly", () => {
    let req = {
      session: {} as any
    };

    let enhancedReq = enhanceRequestWithUser(
      req,
      "cookieName",
      "cookie",
      {} as any,
      {} as any,
      {} as any
    );

    expect(enhancedReq.user).toBeDefined();
    expect(typeof enhancedReq.user.get).toBe("function");
    expect(typeof enhancedReq.user.userInfo).toBe("function");
    expect(typeof enhancedReq.user.fullUserData).toBe("function");
    expect(typeof enhancedReq.user.permissions).toBe("function");
    expect(enhancedReq.user).toMatchSnapshot();
  });
});
