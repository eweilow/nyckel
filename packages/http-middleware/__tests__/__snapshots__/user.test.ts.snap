// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enhanceRequestWithUser should enhance a request correctly 1`] = `
Object {
  "fullUserData": [Function],
  "get": [Function],
  "permissions": [Function],
  "roles": [Function],
  "userInfo": [Function],
}
`;
